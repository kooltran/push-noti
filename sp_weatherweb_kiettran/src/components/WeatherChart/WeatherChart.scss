.chart {
  &-container {
    position: relative;
    overflow-x: scroll;
    height: 40%;
    background-color: $COLOR_WHITE;

    .time-text {
      position: fixed;
      color: $COLOR_BLACK;
      bottom: 0;
      left: 50%;
      transform: translateX(-50%);

      &:before {
        content: "";
        display: inline-block;
        width: 0;
        height: 0;
        border-style: solid;
        border-width: 0 8px 8px 8px;
        border-color: transparent transparent $COLOR_WHITE transparent;
        position: absolute;
        bottom: 20px;
        left: 50%;
        transform: translateX(-50%);
      }

      &:after {
        content: "";
        position: absolute;
        bottom: 20px;
        left: 50%;
        width: 1px;
        height: 150px;
        background: $COLOR_BLACK_FADED;
      }
    }

    .img-sun {
      display: none;
    }

    .icon-moon {
      position: fixed;
      bottom: 25%;
      left: 50%;
      transform: translateX(-50%);
      width: 25px;
      height: 25px;
      background-size: contain;
    }

    .icon-moon {
      background-image: url($ICON_MOON);
    }

    .day-text {
      position: fixed;
      bottom: calc(40% - 60px);
      left: 50%;
      font-size: 24px;
      color: $COLOR_BLACK;
      font-weight: 700;
      transform: translateX(-50%);
    }
  }

  &-title {
    position: fixed;
    bottom: calc(40% - 25px);
    left: 15px;
    width: 100%;

    .blue-title {
      color: $COLOR_LIGHT_BLUE;
      font-size: 18px;
      font-weight: 500;
    }

    .orange-title {
      color: $COLOR_ORANGE;
      font-size: 18px;
      margin-left: 10px;
    }
  }
}
