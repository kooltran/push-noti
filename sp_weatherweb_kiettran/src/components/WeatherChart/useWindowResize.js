import { useState, useLayoutEffect } from "react";

export const useWindowResize = () => {
  const { innerWidth } = window;
  const [chartWidth, setChartWidth] = useState(innerWidth);
  useLayoutEffect(() => {
    const updateChartWidth = (e) => {
      const resizedChartWidth = e && e.currentTarget.innerWidth;
      if (resizedChartWidth) {
        setChartWidth(resizedChartWidth);
      }
    };
    updateChartWidth();
    window.addEventListener("resize", updateChartWidth);
    return () => {
      window.removeEventListener("resize", updateChartWidth);
    };
  }, [chartWidth]);

  return chartWidth;
};
