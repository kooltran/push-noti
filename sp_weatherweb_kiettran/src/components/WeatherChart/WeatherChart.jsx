import React, { useEffect, useState, useCallback } from "react";
import {
  formatTime,
  tideData,
  convertPxToTime,
  getTimeMode,
  getQuadraticCurvePoint,
  convertDataToXY,
} from "../../helpers";
import { useWindowResize } from "./useWindowResize";
import sun from "../../assets/img/icons/sun.svg";

const WeatherChart = () => {
  const chartRef = React.createRef();
  const chartContainerRef = React.createRef();
  const sunRef = React.createRef();
  const chartWidth = useWindowResize();
  const [scrollNumb, setScroll] = useState(0);
  const [tmpPos, setPos] = useState(0.0);
  const [period, setPeriod] = useState(0);
  const [days, setDay] = useState(0);
  const canvasWidth = (tideData.length / 2) * chartWidth;
  const halfChartWidth = chartWidth / 2;
  const pxEachHr = chartWidth / 12;
  const timeFromPx = convertPxToTime(scrollNumb, chartWidth);
  const timeScroll = formatTime(timeFromPx);
  const mode = getTimeMode(timeFromPx).mode;

  const _fillChartBg = ({ start, end, chartHeight, ctx }) => {
    ctx.lineTo(end.x, chartHeight);
    ctx.lineTo(start.x, chartHeight);

    ctx.fillStyle = "#c1e5f7";
    ctx.strokeStyle = "#c1e5f7";
    ctx.fill();
  };

  const _fillText = (start, ctx, dataText) => {
    ctx.font = "14px Comic Sans MS";
    ctx.textAlign = "center";
    ctx.fillStyle = "#333";
    ctx.fillText(`${dataText.waterLevel} m`, start.x, start.y / 2 + 20 + 18);
    ctx.fillText(formatTime(dataText.time), start.x, start.y / 2 + 20 + 35);

    ctx.fillStyle = "rgba(0, 0, 0, 0.1)";
    ctx.fillRect(start.x - 28, start.y / 2 + 20, 60, 44);
  };

  //draw time chart
  const drawTimeChart = useCallback(
    (ctx, chartHeight) => {
      ctx.beginPath();
      ctx.lineWidth = 2;
      ctx.strokeStyle = "#ff8514";

      for (let i = 0; i < tideData.length; i++) {
        ctx.fillStyle = "rgba(0, 0, 0, 0.2)";
        if (i === 0) {
          ctx.fillRect(0, 0, halfChartWidth, chartHeight);
        }

        ctx.moveTo(i * chartWidth + halfChartWidth, chartHeight);
        if (!(i % 2)) {
          ctx.quadraticCurveTo(
            chartWidth * (i + 1),
            -chartHeight + 150,
            chartWidth * (i + 1) + halfChartWidth,
            chartHeight
          );
        } else {
          // fill color for night times
          ctx.fillRect(
            halfChartWidth + chartWidth * i,
            0,
            chartWidth,
            chartHeight
          );
        }
      }

      ctx.stroke();
    },
    [chartWidth, halfChartWidth]
  );

  //draw tide chart
  const drawTideChart = useCallback(
    (ctx, chartHeight) => {
      ctx.beginPath();
      ctx.strokeStyle = "#94d6f7";

      const startPt = (i) => {
        return convertDataToXY(
          chartHeight,
          tideData[i].waterLevel.start,
          tideData[i].time.start,
          pxEachHr
        );
      };

      const endPt = (i) => {
        return convertDataToXY(
          chartHeight,
          tideData[i].waterLevel.end,
          tideData[i].time.end,
          pxEachHr
        );
      };

      const middlePtXFirst = (i) =>
        (endPt(i).x - startPt(i).x) / 2 + startPt(i).x;

      const middlePtYFirst = (i) =>
        (chartHeight - startPt(i).y - (chartHeight - endPt(i).y)) / 2 +
        startPt(i).y;
      const middlePt = (i) => ({
        x: middlePtXFirst(i),
        y: middlePtYFirst(i),
      });

      for (let i = 0; i < tideData.length; i++) {
        _fillText(startPt(i), ctx, {
          waterLevel: tideData[i].waterLevel.start,
          time: tideData[i].time.start,
        });

        if (i === 0) {
          ctx.moveTo(-middlePt(i).x, middlePt(i).y);
          ctx.quadraticCurveTo(
            startPt(i).x,
            startPt(i).y,
            middlePt(i).x,
            middlePt(i).y
          );
          // fill background color for tide chart
          _fillChartBg({ start: startPt(i), end: endPt(i), chartHeight, ctx });
        }

        if (i <= tideData.length - 2) {
          ctx.moveTo(middlePt(i).x, middlePt(i).y);
          ctx.quadraticCurveTo(
            startPt(i + 1).x,
            startPt(i + 1).y,
            middlePt(i + 1).x,
            middlePt(i + 1).y
          );

          // fill background color for tide chart
          _fillChartBg({
            start: middlePt(i),
            end: middlePt(i + 1),
            chartHeight,
            ctx,
          });
        }

        if (i === tideData.length - 1) {
          // fill info of tide chart
          _fillText(endPt(i), ctx, {
            waterLevel: tideData[i].waterLevel.end,
            time: tideData[i].time.end,
          });

          ctx.moveTo(middlePt(i).x, middlePt(i).y);
          ctx.quadraticCurveTo(
            endPt(i).x,
            endPt(i).y,
            endPt(i).x + (endPt(i).x - startPt(i).x),
            startPt(i).y
          );

          // fill background color for tide chart
          _fillChartBg({
            start: middlePt(i),
            end: endPt(i),
            chartHeight,
            ctx,
          });
        }
      }
      ctx.stroke();
    },
    [pxEachHr]
  );

  const drawSun = useCallback(
    (ctx, chartHeight) => {
      //get position of sun base on the TimeChart
      const ps = getQuadraticCurvePoint(
        period * chartWidth + halfChartWidth,
        chartHeight,
        chartWidth * (period + 1),
        -chartHeight + 150,
        (period + 1) * chartWidth + halfChartWidth,
        chartHeight,
        tmpPos
      );

      if (period > 0) {
        setPos((scrollNumb / chartWidth) % 2);
      } else {
        setPos(scrollNumb / chartWidth);
      }

      //draw sun icon
      ctx.drawImage(sunRef.current, ps.x - 10, ps.y - 10, 20, 20);
    },
    [chartWidth, period, halfChartWidth, scrollNumb, sunRef, tmpPos]
  );

  const handleScroll = ({ target }) => {
    const period = Math.floor(scrollNumb / chartWidth);
    setScroll(target.scrollLeft);
    setPeriod(period);
    setDay(Math.floor(timeFromPx / 24));
  };

  const handleCanvas = useCallback(() => {
    const chartHeight =
      chartContainerRef.current && chartContainerRef.current.offsetHeight - 20;
    const c = chartRef.current || {};
    const ctx = chartRef.current && chartRef.current.getContext("2d");

    c.width = canvasWidth;
    c.height = chartHeight;

    if (ctx) {
      drawTideChart(ctx, chartHeight);
      drawTimeChart(ctx, chartHeight);
      drawSun(ctx, chartHeight);
    }
  }, [
    canvasWidth,
    chartContainerRef,
    chartRef,
    drawSun,
    drawTideChart,
    drawTimeChart,
  ]);

  useEffect(() => {
    handleCanvas();
  }, [chartContainerRef, chartWidth, handleCanvas, scrollNumb]);

  return (
    <div
      className="chart-container"
      ref={chartContainerRef}
      onScroll={handleScroll}
    >
      <canvas className="chart-canvas" ref={chartRef} />
      <div className="chart-title">
        <span className="blue-title">Tide</span>
        <span className="orange-title">Sunrise & Sunset</span>
      </div>
      <img ref={sunRef} src={sun} className="img-sun" alt="sun" />
      {mode === "moon" && <div className="icon-moon"></div>}
      <span className="time-text">{timeScroll}</span>
      <div className="day-text">Day {days + 1}</div>
    </div>
  );
};

export default WeatherChart;
