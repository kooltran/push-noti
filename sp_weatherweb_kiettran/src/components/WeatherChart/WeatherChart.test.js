import React from "react";
import renderer from "react-test-renderer";
import WeatherChart from "./WeatherChart";
import Adapter from "enzyme-adapter-react-16";
import { shallow, mount, configure } from "enzyme";

import { renderHook } from "@testing-library/react-hooks";
import { useState } from "react";

configure({ adapter: new Adapter() });

describe("WeatherChart", () => {
  let wrapper;
  // const setScroll = jest.fn();
  // const setPeriod = jest.fn();
  // const setDay = jest.fn();
  // const match = { target: { scrollLeft: 100 } };
  // const handleScroll = jest.fn({ target: { scrollLeft: 100 } });
  // const useStateSpy = jest.spyOn(React, "useState");
  // useStateSpy.mockImplementation((init) => [init, setScroll]);

  const { scrollNumb } = renderHook(() => useState(0));
  it("should render WeatherParams correctly", () => {
    wrapper = renderer.create(<WeatherChart />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("should render handleCanvas function", () => {
    wrapper = mount(<WeatherChart />);
  });

  it("should call the setScroll function when scroll chart container", () => {
    wrapper = shallow(<WeatherChart />);
    wrapper.find(".chart-container").simulate("scroll", { deltaX: 100 });
    // console.log(wrapper.state());
    // expect(scrollNumb).toEqual(0);
  });
});
