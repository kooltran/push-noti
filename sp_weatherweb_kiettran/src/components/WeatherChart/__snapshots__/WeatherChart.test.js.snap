// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherChart should render WeatherParams correctly 1`] = `
<div
  className="chart-container"
  onScroll={[Function]}
>
  <div
    className="chart-title"
  >
    <span
      className="blue-title"
    >
      Tide
    </span>
    <span
      className="orange-title"
    >
      Sunrise & Sunset
    </span>
  </div>
  <canvas
    className="chart-canvas"
  />
  <img
    alt="sun"
    className="img-sun"
    src="sun.svg"
  />
  <span
    className="time-text"
  >
    6:00 am
  </span>
  <div
    className="day-text"
  >
    Day 
    1
  </div>
</div>
`;
